lib/rspec-puppet/monkey_patches.rb

```diff
@@ -315,6 +315,18 @@ def manages_symlinks?
   end
 end
 
+# Puppet 4.8 systemd provider does not recognize Debian 9 (Stretch). It fall
+# back to runit which causes:
+#   Could not find the daemon directory (tested [/etc/sv,/var/lib/service])
+# The Debian package is patched, but the gem 4.8.2 is not.
+# Support has been added in Puppet 4.10.0:
+# https://github.com/puppetlabs/puppet/commit/d5a69fb57c15683873b422cb5e17ef06ca13cea5
+if Puppet::Util::Package.versioncmp(Puppet.version, '4.8.0') >= 0 && Puppet::Util::Package.versioncmp(Puppet.version, '4.10.0') < 0
+  Puppet::Type.type(:service).provide(:systemd).instance_eval do
+    defaultfor :operatingsystem => :debian
```

Comment:
This condition is incorrect, as it will make tests for Debian < 8.0 fail or be inaccurate (those systems don't use systemd).

Also, the condition set in puppet 4.10.x is still wrong and overridden in the debian pacakges to this day.

The correct condition would probably be:
- Monkey-patch all versions that include a systemd provider and offer the defaultfor method for providers
- Add a specific defaultfor to the "debian" provider for debian versions up to 7 (you can limit yourself to 5,6,7)
- Add this defaultfor

Reply:
And on further discussion, @lavagetto pointed that the patch modifies the behavior of the puppet interpreter. That could potentially hide an issue when running puppet from Gem, but I guess everyone just use the patched up puppet debian package.